src/IECore/MeshAlgo.cpp

```diff
@@ -0,0 +1,202 @@
+#include <assert.h>
+#include <algorithm>
+
+#include "OpenEXR/ImathVec.h"
+
+#include "IECore/MeshAlgo.h"
+
+using namespace IECore;
+using namespace Imath;
+
+namespace IECore
+{
+
+namespace MeshAlgo
+{
+
+std::pair<PrimitiveVariable, PrimitiveVariable> calculateTangents(
+	const MeshPrimitive *mesh,
+	const std::string &uvSetName, /* = "st" */
+	bool orthoTangents, /* = true */
+	const std::string &positionPrimVarName /* = "P" */
+)
+{
+	if( mesh->minVerticesPerFace() != 3 || mesh->maxVerticesPerFace() != 3 )
+	{
+		throw InvalidArgumentException( "MeshAlgo::calculateTangents : MeshPrimitive must only contain triangles" );
+	}
+
+	std::string uPrimvarName;
+	std::string vPrimvarName;
+	std::string indicesPrimvarName;
+
+	if( uvSetName == "st" )
+	{
+		uPrimvarName = "s";
+		vPrimvarName = "t";
+	}
+	else
+	{
+		uPrimvarName = uvSetName + "_s";
+		vPrimvarName = uvSetName + "_t";
+	}
+
+	indicesPrimvarName = uvSetName + "Indices";
+
+	const V3fVectorData *positionData = mesh->variableData<V3fVectorData>( positionPrimVarName );
+	if( !positionData )
+	{
+		std::string e = boost::str(
+			boost::format( "MeshAlgo::calculateTangents : MeshPrimitive has no Vertex \"%s\" primitive variable." ) %
+			positionPrimVarName
+		);
+		throw InvalidArgumentException( e );
+	}
+
+	const V3fVectorData::ValueType &points = positionData->readable();
+
+	const IntVectorData *vertsPerFaceData = mesh->verticesPerFace();
+	const IntVectorData::ValueType &vertsPerFace = vertsPerFaceData->readable();
+
+	const IntVectorData *vertIdsData = mesh->vertexIds();
+	const IntVectorData::ValueType &vertIds = vertIdsData->readable();
+
+	const IntVectorData *stIndicesData = mesh->variableData<IntVectorData>( indicesPrimvarName );
+	if( !stIndicesData )
+	{
+		// I'm a little unsure about using the vertIds for the stIndices.
```

Comment:
In the original, this fallback was correct because we knew we were only supporting face varying uvs. I don't think it is correct anymore, given the removal of the FaceVarying check below.